ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java

```diff
@@ -63,4 +66,82 @@ private boolean isValid(AppendEntriesReplyProto reply) {
       return appender.getCallIdComparator().compare(reply.getServerReply().getCallId(), minCallId) >= 0;
     }
   }
+
+  static class HeartbeatBroadcastListener {
+    private final int groupPeerCount;
+    private int ack;
+    private int fail;
+    private volatile boolean done;
+    private final CompletableFuture<Boolean> result;
+
+    HeartbeatBroadcastListener(int groupSize) {
+      this.groupPeerCount = groupSize;
+      this.ack = 0;
+      this.fail = 0;
```

Comment:
When a heartbeat fails (by some network error or other problems), it should retry until it succeeds.  We should not count heartbeat failures.

Reply:
I've removed the fail field while keeps the timeout mechanism. If a broadcast does not complete in a given time, just complete it with value `false`.